packages/aws-cdk/lib/api/hotswap/monitor/logs-monitor.ts

```diff
@@ -0,0 +1,226 @@
+
+  private activity = new Set<string>();
+
+  private logGroups = new Set<string>();
+  private sdk?: ISDK;
+
+  constructor(
+    private readonly printer: IEventPrinter,
+    hotswapTime?: Date,
+  ) {
+    this.startTime = hotswapTime?.getTime() ?? Date.now();
+  }
+
+  public start() {
+    this.active = true;
+    this.scheduleNextTick();
+    return this;
+  }
+
+  public addSDK(sdk: ISDK) {
+    this.sdk = sdk;
+  }
+
+  /**
+   * Adds a CloudWatch log group to read log events from
+   */
+  public addLogGroups(logGroups: string[]) {
+    logGroups.forEach(group => {
+      if (!this.logGroups.has(group)) {
+        this.logGroups.add(group);
+      }
+    });
+  }
+
+  public async stop() {
+    this.active = false;
+    if (this.tickTimer) {
+      clearTimeout(this.tickTimer);
+    }
+  }
+
+  private scheduleNextTick() {
+    if (!this.active) {
+      return;
+    }
+    this.tickTimer = setTimeout(() => void (this.tick()), this.printer.updateSleep);
+  }
+
+  private async tick() {
+    if (!this.active) {
+      return;
+    }
+    try {
+      this.readPromise = this.readNewEvents();
+      await this.readPromise;
+      this.readPromise = undefined;
+
+      if (!this.active) { return; }
+
+    } catch (e) {
+      error('Error occurred while monitoring logs: %s', e);
+    }
+    this.scheduleNextTick();
+
+  }
+
+  /**
+   * Reads all new log events from a set of CloudWatch Log Groups
+   * in parallel
+   */
+  private async readNewEvents(): Promise<void[]> {
+    const promises: Array<Promise<void>> = [];
+    for (let group of this.logGroups) {
+      promises.push(this.readEventsFromLogGroup(group));
+    }
+    return Promise.all(promises);
+  }
+
+  /**
+   * Reads all new log events from a CloudWatch Log Group
+   * starting when the hotswap was triggered
+   *
+   * Only prints out events that have not been printed already
+   */
+  private async readEventsFromLogGroup(logGroupName: string): Promise<void> {
+    // this should not be possible
+    if (!this.sdk) {
+      return;
+    }
+    try {
+      let finished = false;
+      let nextToken: string | undefined;
+      while (!finished) {
+        const response = await this.sdk!.cloudwatchLogs().filterLogEvents({
+          logGroupName: logGroupName,
+          nextToken,
+          startTime: this.startTime,
+        }).promise();
+        const eventPage = response.events ?? [];
+
+
+        for (const event of eventPage) {
+          if (event.eventId && this.activity.has(event.eventId)) {
+          } else {
+            if (event.eventId) this.activity.add(event.eventId);
+            if (event.message) {
+              const ingestionTime: Date = event.ingestionTime ?
+                new Date(event.ingestionTime) : new Date();
+              this.printer.print({
+                message: event.message,
+                logGroup: logGroupName,
+                logStream: event.logStreamName,
+                ingestionTime,
+              });
```

Comment:
I'm not sure I agree the method `readEventsFromLogGroup()` should also be doing the printing.

How about you make `readEventsFromLogGroup()` return `Promise<CloudWatchEvent[]>`, and then `tick()` will actually do the printing?

Reply:
Updated